tests/Progress/SnapshotTest.php

```diff
@@ -359,12 +361,223 @@ public function of_activity_runs_the_above_for_all_activity_instances()
+        $activityInstance = factory(ActivityInstance::class)->create(['activity_id' => $activity->id]);
+
+        $module1Progress = $this->createModuleInstanceProgress($moduleInstance->id, true, true, true, true, 100);
+        $module1Evaluation = $this->createModuleEvaluationInstance(true, true, true, true, 100);
+
+        $moduleInstanceEvaluator = $this->prophesize(ModuleInstanceEvaluator::class);
+        $this->useEvaluationForModuleInstance($moduleInstanceEvaluator, $activityInstance, $moduleInstance, $module1Evaluation);
+        \BristolSU\Support\ModuleInstance\Facade\ModuleInstanceEvaluator::swap($moduleInstanceEvaluator->reveal());
+
+        $progressUpdateRepository = $this->prophesize(ProgressUpdateRepository::class);
+
+        $progressUpdateRepository->hasChanged(
+            $activityInstance->id,
+            'my-caller',
+            Argument::that(function($progress) use ($module1Progress) {
+                return $progress->getModules()[0] == $module1Progress;
+        }))->willReturn(false);
+
+        $snapshot = new Snapshot($progressUpdateRepository->reveal());
+
+        $this->assertNull(
+            $snapshot->ofUpdateToActivityInstance($activityInstance, 'my-caller')
+        );
+    }
+
+    /** @test */
+    public function ofUpdatesToActivityInstance_returns_progress_if_the_progress_has_changed(){
+        Carbon::setTestNow(Carbon::now());
+        $activity = factory(Activity::class)->create();
+        $moduleInstance = factory(ModuleInstance::class)->create(['activity_id' => $activity->id]);
+        $activityInstance = factory(ActivityInstance::class)->create(['activity_id' => $activity->id]);
+
+        $module1Progress = $this->createModuleInstanceProgress($moduleInstance->id, true, true, false, true, 50);
+
+        $module1Evaluation = $this->createModuleEvaluationInstance(true, true, true, true, 100);
+
+        $moduleInstanceEvaluator = $this->prophesize(ModuleInstanceEvaluator::class);
+        $this->useEvaluationForModuleInstance($moduleInstanceEvaluator, $activityInstance, $moduleInstance, $module1Evaluation);
+        \BristolSU\Support\ModuleInstance\Facade\ModuleInstanceEvaluator::swap($moduleInstanceEvaluator->reveal());
+
+        $progressUpdateRepository = $this->prophesize(ProgressUpdateRepository::class);
+
+        $progressUpdateRepository->hasChanged(
+            $activityInstance->id,
+            'my-caller',
+            Argument::that(function($progress) use ($module1Progress) {
+                return $progress->getModules()[0] != $module1Progress;
+            }))->willReturn(true);
+
+        $progressUpdateRepository->saveChanges(
+            $activityInstance->id,
+            'my-caller',
+            Argument::that(function($progress) use ($module1Progress) {
+                return $progress->getModules()[0] != $module1Progress;
+            })
+        );
+
+        $snapshot = new Snapshot($progressUpdateRepository->reveal());
+
+        $this->assertNotNull(
```

Comment:
Even better, could you check it returns a Progress instance with values as you'd expect? e.g. check complete/visible/mandatory are true etc?